Lib/glyphsLib/glyphdata.py

```diff
@@ -14,83 +14,181 @@
+)
+
+import collections
+import os
+import re
 import unicodedata
+import xml.etree.ElementTree
 
-NARROW_PYTHON_BUILD = sys.maxunicode < 0x10FFFF
+import fontTools.agl
 
+import glyphsLib
 
 # FIXME: (jany) Shouldn't this be the class GSGlyphInfo?
-Glyph = namedtuple("Glyph", "name,production_name,unicode,category,subCategory")
+Glyph = collections.namedtuple(
+    "Glyph", "name, production_name, unicode, category, subCategory"
+)
 
+GLYPHSDATA = None
 
-def get_glyph(glyph_name, data=glyphdata_generated):
+
+class GlyphsData:
+    __slots__ = ["names", "alternative_names", "production_names"]
+
+    def __init__(
+        self, name_mapping, alt_name_mapping, production_name_mapping
+    ):
+        self.names = name_mapping
+        self.alternative_names = alt_name_mapping
+        self.production_names = production_name_mapping
+
+    @classmethod
+    def from_files(cls, *glyphdata_files):
+        name_mapping = {}
+        alt_name_mapping = {}
+        production_name_mapping = {}
+
+        for glyphdata_file in glyphdata_files:
+            glyph_data = xml.etree.ElementTree.parse(glyphdata_file).getroot()
+            for glyph in glyph_data:
+                glyph_name = glyph.attrib["name"]
+                glyph_name_alternatives = glyph.attrib.get("altNames")
+                glyph_name_production = glyph.attrib.get("production")
+
+                name_mapping[glyph_name] = glyph.attrib
+                if glyph_name_alternatives:
+                    alternatives = glyph_name_alternatives.replace(
+                        " ", ""
+                    ).split(",")
+                    for glyph_name_alternative in alternatives:
+                        alt_name_mapping[glyph_name_alternative] = glyph.attrib
+                if glyph_name_production:
+                    production_name_mapping[
+                        glyph_name_production
+                    ] = glyph.attrib
+
+        return cls(name_mapping, alt_name_mapping, production_name_mapping)
+
+
+def get_glyph(glyph_name):
     """Return a named tuple (Glyph) containing information derived from a glyph
     name akin to GSGlyphInfo.
 
-    The information is derived from an included copy of GlyphsData.xml,
+    The information is derived from an included copy of GlyphData.xml,
     going purely by the glyph name.
     """
+    global GLYPHSDATA
+
+    if GLYPHSDATA is None:
+        GLYPHSDATA = GlyphsData.from_files(
+            os.path.join(
+                os.path.dirname(glyphsLib.__file__), "data", "GlyphData.xml"
+            ),
+            os.path.join(
+                os.path.dirname(glyphsLib.__file__),
+                "data",
+                "GlyphData_Ideographs.xml",
+            ),
+        )
 
-    # First, get the base name of the glyph. .notdef and .null are exceptions.
-    # Periods denote glyph variants as per the AGLFN convention, which should
-    # be in the same category as their base glyph.
-    if glyph_name in (".notdef", ".null"):
```

Comment:
do we not need to special-case '.notdef' and '.null' any more here?

Reply:
Both are part of the XML data.